Formula/macaulay2.rb

```diff
@@ -0,0 +1,56 @@
+class Macaulay2 < Formula
+  @name = "M2"
+  desc "Software system for algebraic geometry research"
+  homepage "http://macaulay2.com"
+  url "https://github.com/Macaulay2/M2.git", :using => :git, :branch => "release-1.16"
+  version "1.16"
+  license "GPL-3.0"
+
+  head do
+    url "https://github.com/Macaulay2/M2.git", :using => :git, :branch => "development"
+  end
+
+  # autoconf, automake, and libtool are required to build
+  # libraries not available via brew, such as factory
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "bison" => :build
+  depends_on "cmake" => :build
+  depends_on "libtool" => :build
+  depends_on "ninja" => :build
+  depends_on "pkg-config" => :build
+
+  # libraries available via brew
+  depends_on "bdw-gc"
+  depends_on "boost"
+  depends_on "cddlib"
+  depends_on "eigen"
+  depends_on "flint"
+  depends_on "gdbm"
+  depends_on "glpk"
+  depends_on "gmp"
+  depends_on "libatomic_ops"
+  depends_on "libomp"
+  depends_on "libxml2"
+  depends_on "mpfr"
+  depends_on "ntl"
+  depends_on "openblas"
+  depends_on "readline"
+  depends_on "tbb"
+
+  def install
+    build_path = "M2/BUILD/brew"
+    extra_args = ["-GNinja", "-DUSING_MPIR=OFF", "-DBLA_VENDOR=OpenBLAS", "-DCMAKE_INSTALL_LIBDIR=lib"]
+
+    system "cmake", "-SM2", "-B#{build_path}", *std_cmake_args, *extra_args
+    system "cmake", "--build", build_path, "--target", "build-libraries", "build-programs"
+    system "cmake", "--build", build_path, "--target", "M2-core", "M2-emacs"
+    system "cmake", "--build", build_path, "--target", "install-packages"
+    system "cmake", "--install", build_path
+  end
+
+  test do
+    system "#{bin}/M2", "--version"
+    system "#{bin}/M2", "--check", "1", "-e", "exit 0"
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
The `M2 --check 1` will run a series of [internal basic tests](https://github.com/Macaulay2/M2/tree/master/M2/Macaulay2/m2/basictests) exactly for this purpose. The source script for the tests, including their assertions, are hardcoded in the binary.

Reply:
Here's a sample output:
```m2
[mahrud@noether build]$ M2 --check 1 -e "exit 0"
-- testing Macaulay2/Core/basictests/A01.m2
-- testing Macaulay2/Core/basictests/A02.m2
-- testing Macaulay2/Core/basictests/A03.m2
-- testing Macaulay2/Core/basictests/A04.m2
-- testing Macaulay2/Core/basictests/A05.m2
-- testing Macaulay2/Core/basictests/A06.m2
-- testing Macaulay2/Core/basictests/A07.m2
-- testing Macaulay2/Core/basictests/A08.m2
-- testing Macaulay2/Core/basictests/A10.m2
-- testing Macaulay2/Core/basictests/A44.m2
-- testing Macaulay2/Core/basictests/A66.m2
-- testing Macaulay2/Core/basictests/A77.m2
-- testing Macaulay2/Core/basictests/D01.m2
-- testing Macaulay2/Core/basictests/L00.m2
-- testing Macaulay2/Core/basictests/L00a.m2
-- testing Macaulay2/Core/basictests/L01.m2
-- testing Macaulay2/Core/basictests/L02.m2
-- testing Macaulay2/Core/basictests/W01.m2
-- testing Macaulay2/Core/basictests/W02-threads.m2
-- testing Macaulay2/Core/basictests/ascii.m2
-- testing Macaulay2/Core/basictests/bigreal.m2
-- testing Macaulay2/Core/basictests/hashcodes.m2
-- testing Macaulay2/Core/basictests/regex.m2
-- testing Macaulay2/Core/basictests/version.m2
Macaulay2, version 1.16.0.1
--loading configuration for package "FourTiTwo" from file /home/mahrud/.Macaulay2/init-FourTiTwo.m2
--loading configuration for package "Topcom" from file /home/mahrud/.Macaulay2/init-Topcom.m2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, MinimalPrimes, PrimaryDecomposition,
               ReesAlgebra, TangentCone, Truncations
```

Reply:
Is that something users will regularly run?

Reply:
They probably wouldn't run those tests, but the tests verify that the basics of the interpreter are working.